internal/input/orderbooks.go

```diff
@@ -0,0 +1,258 @@
+
+	for _, converted := range allConverted {
+		if _, exists := o.SeenMarketHashes[converted.Market.ID]; !exists {
+			o.SeenMarketHashes[converted.Market.ID] = true
+			marshalledMarket, err := json.Marshal(converted.Market)
+			if err != nil {
+				errorMsg := fmt.Sprintf("error json marshalling market for offer: %d", converted.Offer.HorizonID)
+				if o.Strict {
+					o.Logger.Fatal(errorMsg, err)
+				} else {
+					o.Logger.Warning(errorMsg, err)
+					continue
+				}
+			} else {
+				o.Markets = append(o.Markets, marshalledMarket)
+			}
+		}
+
+		if _, exists := o.SeenAccountHashes[converted.Account.ID]; !exists {
+			o.SeenAccountHashes[converted.Account.ID] = true
+			marshalledAccount, err := json.Marshal(converted.Account)
+			if err != nil {
+				errorMsg := fmt.Sprintf("error json marshalling account for offer: %d", converted.Offer.HorizonID)
+				if o.Strict {
+					o.Logger.Fatal(errorMsg, err)
+				} else {
+					o.Logger.Warning(errorMsg, err)
+					continue
+				}
+			} else {
+				o.Accounts = append(o.Accounts, marshalledAccount)
+			}
+		}
+
+		if _, exists := o.SeenOfferHashes[converted.Offer.DimOfferID]; !exists {
+			o.SeenOfferHashes[converted.Offer.DimOfferID] = true
+			marshalledOffer, err := json.Marshal(converted.Offer)
+			if err != nil {
+				errorMsg := fmt.Sprintf("error json marshalling offer: %d", converted.Offer.HorizonID)
+				if o.Strict {
+					o.Logger.Fatal(errorMsg, err)
+				} else {
+					o.Logger.Warning(errorMsg, err)
+					continue
+				}
+			} else {
+				o.Offers = append(o.Offers, marshalledOffer)
+			}
+		}
+
+		marshalledEvent, err := json.Marshal(converted.Event)
+		if err != nil {
+			errorMsg := fmt.Sprintf("error json marshalling event for offer: %d", converted.Offer.HorizonID)
+			if o.Strict {
+				o.Logger.Fatal(errorMsg, err)
+			} else {
+				o.Logger.Warning(errorMsg, err)
+				continue
+			}
+		} else {
+			o.Events = append(o.Events, marshalledEvent)
+		}
+	}
+}
+
+func GetOfferChanges(core *ledgerbackend.CaptiveStellarCore, firstSeq, nextSeq uint32) (*ingestio.LedgerEntryChangeCache, error) {
+	offChanges := ingestio.NewLedgerEntryChangeCache()
+
+	for seq := firstSeq; seq <= nextSeq; {
+		changeReader, err := ingestio.NewLedgerChangeReader(core, password, seq)
+		if err != nil {
+			return nil, fmt.Errorf(fmt.Sprintf("unable to create change reader for ledger %d: ", seq), err)
+		}
+
+		err = addLedgerChangesToCache(changeReader, nil, offChanges, nil)
+		if err != nil {
+			return nil, fmt.Errorf(fmt.Sprintf("unable to read changes from ledger %d: ", seq), err)
+		}
+
+		changeReader.Close()
+		seq++
+	}
+
+	return offChanges, nil
+}
+
+func exportOrderbookBatch(batchStart, batchEnd uint32, core *ledgerbackend.CaptiveStellarCore, orderbookChan chan OrderbookBatch, startOrderbook []ingestio.Change, logger *log.Entry) {
+	batchMap := make(map[uint32][]ingestio.Change)
+	batchMap[batchStart] = make([]ingestio.Change, len(startOrderbook))
+	copy(batchMap[batchStart], startOrderbook)
```

Comment:
Why the copy over assigment? My understanding is `make(map...)` allows for dynamic sizing of values, so you shouldn't need to make then copy. Could be totally off, since I can't find direct documentation.

i.e., `batchMap[batchStart] = startOrderbook`

Reply:
The orderbook gets modified later on as new changes are applied to the orderbook. If we don't copy values, then the value in the map ends up changing as well, since its just a reference.